libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Same.hs

```diff
@@ -0,0 +1,535 @@
+sameWithDependency (SomeM labx actx x1 x2 : more) =
+  case actx x1 x2 of
+    [] -> sameWithDependency more
+    ansx -> extendLabel (labx ++ " ") ansx ++ sameWithDependency more
+
+instance Reflect era => Same era (ShelleyLedgerExamples era) where
+  same
+    proof
+    (ShelleyLedgerExamples (Block' h1 x1 _) hh1 tx1 ap1 rc1 re1 ne1 cd1 ctx1)
+    (ShelleyLedgerExamples (Block' h2 x2 _) hh2 tx2 ap2 rc2 re2 ne2 cd2 ctx2) =
+      sameWithDependency
+        [ SomeM "Tx" (sameTx proof) tx1 tx2,
+          SomeM "TxSeq" (sameTxSeq proof) x1 x2
+        ]
+        ++ [ ("BlockHeader", if h1 == h2 then Nothing else Just ("UnequalBlockHeader")),
+             ("HashHeader", if hh1 == hh2 then Nothing else Just ("UnequalHashHeader")),
+             ("ApplyTxError", sameLedgerFail proof ap1 ap2),
+             ("RewardsCredentials", eqByShow rc1 rc2)
+           ]
+        ++ extendLabel "Result " (same proof re1 re2)
+        ++ extendLabel "NewEpochState " (same proof ne1 ne2)
+        ++ [ ("ChainDepState", eqByShow cd1 cd2),
+             ("TranslationContext", sameTransCtx proof ctx1 ctx2)
+           ]
+
+instance Same era (ShelleyResultExamples era) where
+  same proof (ShelleyResultExamples pp1 ppup1 pdist1 myop1 gen1) (ShelleyResultExamples pp2 ppup2 pdist2 myop2 gen2) =
+    [ ("PParams", samePParams proof pp1 pp2),
+      ( "ProposedPPUpdates",
+        case proof of
+          Shelley _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+          Allegra _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+          Mary _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+          Alonzo _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+          Babbage _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+          Conway _ ->
+            case mapdiffs (getmap ppup1) (getmap ppup2) of
+              [] -> Nothing
+              xs -> Just (ppString (show xs))
+      ),
+      ("poolDistr", eqByShow pdist1 pdist2),
+      ("NonMyopicRewards", eqByShow myop1 myop2),
+      ("ShelleyGenesis", eqByShow gen1 gen2)
+    ]
+    where
+      getmap (ProposedPPUpdates x) = x
+
+-- =========================================================================
+-- Functions like 'same' from the 'Same' class, but which apply to type families
+-- We cannot make them Same instances because they are type families.
+-- We also can avoid all extra constraints by pattern matching against all current Proofs.
+
+samePParams :: Proof era -> Core.PParams era -> Core.PParams era -> Maybe PDoc
```

Comment:
No, that is not the case. Specialize only works if the context needed in every case is a real class (like Eq or Show)
In order to get it to typecheck one usually has to type apply the call, to be concrete about what that class is.
For example we may say  "specialize @Eq proof x y"  Many times the constraints differ from Era to Era,  or the constraint is some compound constrant like  \ z -> (Eq z, Show z)  which we can't really express. 

I wish we had a better more flexible approach, but I have yet to find it.

Reply:
I see, the `constraint` type parameter in `specialize` needs to be instantiated to a concrete class. I wish there were anonymous lambdas at the type level. At this point it can only be done with named type synonym constraints, eg.
```
type  Combine z = (Eq z, Show z)
```

Even if it would work with specialize, I don't think it is worth littering the code with bunch of random type synonyms.

Don't worry about this, I was just curious.